### PR TITLE
feat: getAllValues, getErrorMessagesByField + fix: useSubmission edge case + feat: multiple checkboxes

### LINKED ISSUE #6: useSubmissionState possible edge case
Currently in `useSubmissionState.js` you have something like:

```javascript
(
  onSubmitRef.current({
    allValues,
    registeredValues,
  }) || Promise.resolve()
)
  .then(() => { ...
```

The problem is that if the `onSubmit` function returns a non-falsy non-`Promise` value (which might happen, who knows), the `.then` will kill the code.

I propose changing it to:

```javascript
(
  Promise.resolve(
    onSubmitRef.current({
      allValues,
      registeredValues,
    })
  )
)
  .then(() => { ...
```

You're ignoring the value in the code that follows anyway, so it won't change functionality, but it will plug the possible edge case error.

### LINKED ISSUE #11: More form-global information
Now that I can use OneForm in my project (I set up a local npm registry to use the latest version of the code until you update it on npm), I ran into a small issue: the availability of information about the form without going through fields one by one. While developing the form, I find it useful to have a small component just print out the form state so I can see that things are wired correctly (especially while creating form components that might not communicate well with OneForm yet).

For the form data itself I got away with using a context that gets updated with the `onChange` function on the OneForm component (even has a nice debounce happening so that's good, but it would have been a bit more work if I wasn't using this pattern in other context providers and had a ready-made hook for making it). It seems like a pity to have the information in internal state and not expose it to users (who might need it for more complex use-cases, or just for debugging purposes).

For the errors, it seems way more complicated, and is an even more common use-case (showing a list of all the errors in one place; though it's probably bad UX, people still use it). You have a hook for it (`useErrorMessagesState`), but it's not exported out of the library, and even when it is, it flattens the error object, so we lose the field => errors object and only keep all the errors and functions to get the errors by field name.

### LINKED ISSUE #12: Multiple checkboxes
There's a use-case for multiple checkboxes (using both `isMultipleElement` and `isCheckboxElement`): [see an example here that it's a valid use-case](https://primefaces.org/primereact/showcase/#/checkbox) (taken from the component library I'm using). The resulting effect is somewhat like a multi-select, but with individual checkboxes (I've actually used such a construct a few times in past projects)

Pretty much this would boil down to:
```html
<Field isCheckboxElement isMultipleElement><Checkbox name="languagesSpoken" value="en" /></Field>
<Field isCheckboxElement isMultipleElement><Checkbox name="languagesSpoken" value="ro" /></Field>
<Field isCheckboxElement isMultipleElement><Checkbox name="languagesSpoken" value="pt" /></Field>
```

I've encountered 2 issues with implementing this use-case, both in `useField.js`'s return statement:
1. `value`: if a value isn't set in the default form values for this field, it will actually overwrite the values in the child checkboxes with `[]`, so changing the expression to something like:
```typescript
{ value: isMultipleElement && value === '' && (inputValue == null || inputValue === '') ? [] : inputValue || value }
```
2. `isChecked`: the conditions don't work for checkbox+multiple. I suggest using:
```typescript
{
isChecked: isCheckboxElement
  ? (isMultipleElement
      ? value?.includes(inputValue)
      : Boolean(value))
  : value === inputValue
      ? value
      : false
}
```

I've looked through the stories and seen the `MultiPicker` one, but tbh that's a bit convoluted when it could be easily fixed to "just work" and look more like one would implement radio buttons.

### PR DESCRIPTION
**Pull request by souldreamer**
fixes #6 - as discussed in issue
fixes #11 - adds `getAllValues` and `getErrorMessagesByField` functions to the return of `useFormSubmission`
fixes #12 - adds possiblity of multiple checkboxes (without making a "multiple checkbox" input component)

#### Notes:
* hope I got the formatting right btw)
* Not sure about naming the errors grouped by field function `getErrorMessagesByField`, you might want another name (let me know though so I can change the typings)
* commit c794f26 is if you agree with the idea proposed in #12 
* will probably need extra tests for at least `getAllValues` and `getErrorMessagesByField`

Sorry for the multi-PR, but forking the project again for each of them and doing a `yarn install` every time might just have killed me, it takes way way way too long (and I've also been spoiled by yarn 3) 😅

### CODE REVIEW

**Comment by Sawtaytoes** [comment]
Thanks for this PR!

## Yarn v1
Most of the slowness when running `yarn` v1 is actually from Cypress. I originally didn't have it part of this project but added it to reduce the overhead. Since I'm more versed in Lerna and Yarn workspaces, I could move the Cypress and Storybook runners back into another project.

I could always upgrade this to Yarn v3 in the future as well.

## Values
Because we have `allFields` and `registeredFields` in `onSubmit`, we should probably also have `getRegisteredFieldValues` as well. That means there's one less thing for people to keep track of outside `OneForm`.\

I don't like people having to know the internals of OneForm to use it, but I think the concept of registered fields is crucial to using OneForm.

## Error Messages
I like your idea of having a separate function for error messages by field rather than just the "get all error messages" I put before; although, if it's an object, we could simply have them run `Object.values(errorMessages)` for the same behavior.

I'm thinking it might work even better as an array of objects which would allow you to easily loop 'n group such as:
```js
[
  {
    errorMessages: [],
    fieldName: '',
  }
]
```

At that point, it would be `errors`, not `errorMessages`. At that point, we'd name it `getErrors`.

I'm also thinking we should remove `getAllErrorMessages`. I added it recently, and we can put a deprecation warning in there for anyone using it.

## Multiple Checkboxes
I'm thinking this change is unnecessary since multiple checkboxes already works natively. You can verify by going to this Storybook link: http://localhost:6006/?path=/story/fields-field--multi-picker-component.

I think you found a flaw where I allowed 2 booleans instead of defining a finite state machine. I'll post in that issue with more information.

File: packages/react/src/useSubmissionState.js [hunk 1]
```diff
@@ -190,17 +190,14 @@ const useSubmissionState = (
...
-            .resolve()
-          )
+          Promise
+            .resolve(
+              onSubmitRef
+              .current({
+                allValues,
+                registeredValues,
+              })
+            )
```
Comment (Sawtaytoes) [inline packages/react/src/useSubmissionState.js:200]: 👍

File: packages/react/src/useFormSubmission.js [hunk 2]
```diff
@@ -172,7 +174,9 @@ const useFormSubmission = (
 
   return {
     formChangeState,
+    getAllFieldValues,
     getErrorMessages,
+    getErrorMessagesByField,
```
Comment (Sawtaytoes) [inline packages/react/src/useFormSubmission.js:179]: I'd like to also see `getRegisteredFieldValues` and the deprecation of `getErrorMessages` since `getErrorMessagesByField` is the correct implementation I meant to do in the first place.

I'd also like to see `getErrorMessagesByField` changed to `getErrors` and converted to an array of objects with an `errorMessages` and `fieldName` prop. Makes it easier for consumers to loop and group it. That's going to be the most-likely use case.

File: packages/react/src/useField.js [hunk 3]
```diff
@@ -234,7 +234,14 @@ const useField = ({
...
       ? (
-        Boolean(
+        isMultipleElement
+        ? (
+          value
+          .includes(
+            inputValue
+          )
+        )
+        : Boolean(
```
Comment (Sawtaytoes) [inline packages/react/src/useField.js:244]: 👍 Very simple workaround in this particular use case, but doesn't scale if we had more input types.

Since input types are defined by HTML, I think we're fine. `type="checkbox"` and `multiple` are the only ones that exist today.

File: packages/react/src/useField.js [hunk 4]
```diff
@@ -256,6 +263,16 @@ const useField = ({
           value
           === ''
         )
+        && (
+          (
+            inputValue
+            == null
+          )
```
Comment (Sawtaytoes) [inline packages/react/src/useField.js:270]: Are we using `== null` instead of `=== null` to accommodate `undefined`?

If so, I prefer being explicit and adding `=== undefined` as well. I know it's an extra check, but it makes our intention clear.

I never use `==`, so I had to run it through some possible edge-case tests to ensure it didn't break:

image

To make it easier on other devs, it's better to always be explicit about this kinda stuff.

File: packages/react/src/useField.js [hunk 5]
```diff
@@ -256,6 +263,16 @@ const useField = ({
...
+          || (
+            inputValue
+            === ''
+          )
```
Comment (Sawtaytoes) [inline packages/react/src/useField.js:274]: I think the `''` check should come before the `null`-`undefined` check.

I say that because we're doing `value === ''` the line above. When I read this in the diff, it was difficult to understand the relation between the two. Having the `''` checks ordered first would make it more clear what's going on without having to think about it.

File: packages/react/src/useSubmissionState.js (hunk 1 above)
Comment (Sawtaytoes) [reply packages/react/src/useSubmissionState.js:200]: Since you asked about formatting, there's an extra tab where `resolve` is listed (that entire section). I like to have `.` on the same line so spacing defines scope rather than parens.

Honestly, I can go back and fix these little things on my own time if I feel it's necessary. I'd rather you feel more comfortable contributing to the project and not have to worry about formatting 😉.

---

Both Prettier and ESLint do the "tab on `.`" by default, so it's common practice even though it adds a ton of unnecessary tabbing and makes vertical code harder to follow. I also use line-wrap in my text editor which makes `.` tabbing even more problematic.

Eventually, I'll have a Prettier plugin to adjust formatting the way I like it, but that's not gonna be done for a while. I forked Prettier a while ago and was nearly done with it, but I need to release it as a plugin for a production release.

File: packages/react/src/useField.js (hunk 4 above)
Comment (souldreamer) [reply packages/react/src/useField.js:270]: I always use `== null` for `null | undefined` checks and `=== null` for just `null` checks (which tbh are pretty rare, in 99% of cases you want to check for both). Always seemed pretty clear to me (and to many others, as it's the one exception to the `===` linter checks allowed 😁).

However, I'm trying to keep to your coding style, so will happily make the change.

File: packages/react/src/useSubmissionState.js (hunk 1 above)
Comment (souldreamer) [reply packages/react/src/useSubmissionState.js:200]: Literally half of the time I spent on this PR was fighting with vscode about spacing and deactivating formatters 🤣 (and even then it liked adding extra spacing which I kept fighting to undo and then it would be redone when writing the next line and so on). Totally missed that extra indent, I'd noticed how your code style continued the next line on the same indent level. For more horizontal programming styles, an indent on the next line makes sense, because you have things like:
```typescript
Promise.resolve(...)
  .then(...)
  .catch(...)
```
so it allows you to quickly notice that it's a function chain block.

A prettier plugin would be awesome! Then I could have a git hook to use prettier with my own settings when downloading the code and with yours when uploading 😁.

File: packages/react/src/useFormSubmission.js (hunk 2 above)
Comment (souldreamer) [reply packages/react/src/useFormSubmission.js:179]: Done, though doing an `Object.entries(...).map(...)` on an object is easier than reconstituting the object from the array :D

Comment (souldreamer) [reply packages/react/src/useFormSubmission.js:179]: How would you do the deprecation? Right now, I've deprecated it in Typescript (but that's nothing more than a warning in the editor while trying to use it), but couldn't find how you deprecate things. I usually go for `console.warn`, but you might think that excessive.

Comment (Sawtaytoes) [reply packages/react/src/useFormSubmission.js]: I have a deprecation utility in one of my micro-service libraries, but it's Node.js-only.

I don't mind the TypeScript deprecation. I can also deprecate or remove it in the docs.

Since this is a brand new feature, I honestly don't think anyone was using it; although, who knows! 🤷‍♂️. Just trying to be safe as a library author.

File: packages/react/src/useField.js (hunk 4 above)
Comment (Sawtaytoes) [reply packages/react/src/useField.js]: I'm surprised the linter allows it, and I can I see why it exists so long as it's only going to catch `null` and `undefined` and there's no edge case.

I'd have to ask around to see if others know this syntax too. The first time I saw `!=` instead of `!==` used was at my last job. It's one of those "you have to really know JS to understand this syntax" type of issues and could be something where someone does this without realizing the impact:
```diff
- == null
+ === null
```

It's also really difficult to spot in code reviews in general.

It's not my coding style so much as I'm trying to make the intent obvious.

If I found a good argument that said "everyone in JS knows this thing about `== null` and `!= null`", then I'd change my ways. As it stands, I'm unaware how common this is considering I only stumbled upon it because they didn't use `===` anywhere at my last job. Might just be me.

**Comment by Sawtaytoes** [comment]
At this point, I think you've made all the changes and everything's looking really good.

You cool if I merge it?

**Comment by Sawtaytoes** [comment]
Thanks for all your hard work @souldreamer! These changes are really nice to have!

Glad to have you taking a look at the library and helping out!
